src/Components/WebAssembly/WebAssembly.Authentication/src/RemoteAuthenticatorViewCore.cs

```diff
@@ -329,14 +330,37 @@ private string GetReturnUrl(TAuthenticationState state, string defaultReturnUrl
             return state.ReturnUrl;
         }
 
-        var fromQuery = QueryStringHelper.GetParameter(new Uri(Navigation.Uri).Query, "returnUrl");
+        var fromQuery = GetParameterFromQueryString("returnUrl");
         if (!string.IsNullOrWhiteSpace(fromQuery) && !fromQuery.StartsWith(Navigation.BaseUri, StringComparison.Ordinal))
         {
             // This is an extra check to prevent open redirects.
             throw new InvalidOperationException("Invalid return url. The return url needs to have the same origin as the current page.");
         }
 
         return fromQuery ?? defaultReturnUrl ?? Navigation.BaseUri;
+
+    }
+
+    private string GetParameterFromQueryString(ReadOnlySpan<char> parameterName)
+    {
+        var url = Navigation.Uri;
+        ReadOnlyMemory<char> query = default;
+        var queryStartPos = url.IndexOf('?');
+        if (queryStartPos >= 0)
+        {
+            var queryEndPos = url.IndexOf('#', queryStartPos);
+            query = url.AsMemory(queryStartPos..(queryEndPos < 0 ? url.Length : queryEndPos));
+        }
+
+        foreach (var parameter in new QueryStringEnumerable(query))
+        {
+            var decodedName = parameter.DecodeName().Span;
+            if (MemoryExtensions.Equals(parameterName, decodedName, StringComparison.OrdinalIgnoreCase))
```

Comment:
If you think there's no legitimate case for encoding the `returnUrl` parameter **name** in a special way, then you could avoid an allocation by not decoding it before doing this comparison.

However this is not exactly perf-critical code (it doesn't run super-frequently as far as I can guess) so I think it's fine either way.

Reply:
Oops, no, forget that. If the value doesn't need decoding, it won't cause an allocation anyway as `DecodeName` is smart enough to return the original value directly.